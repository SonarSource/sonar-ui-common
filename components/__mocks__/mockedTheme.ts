/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

// IMPORTANT: any change in this file requires restart of the dev server
const grid = 8;

export const mockedTheme = {
  colors: {
    blue: '#4b9fd5',
    veryLightBlue: '#f2faff',
    lightBlue: '#cae3f2',
    darkBlue: '#236a97',
    veryDarkBlue: '#0E516F',
    green: '#00aa00',
    lightGreen: '#b0d513',
    veryLightGreen: '#f5f9fc',
    yellow: '#eabe06',
    orange: '#ed7d20',
    red: '#d4333f',
    purple: '#9139d4',

    gray94: '#efefef',
    gray80: '#cdcdcd',
    gray71: '#b4b4b4',
    gray67: '#aaa',
    gray60: '#999',
    gray40: '#404040',

    transparentWhite: 'rgba(255,255,255,0.62)',
    transparentGray: 'rgba(200, 200, 200, 0.5)',

    disableGrayText: '#bbb',
    disableGrayBorder: '#ddd',
    disableGrayBg: '#ebebeb',

    barBackgroundColor: '#f3f3f3',
    barBorderColor: '#e6e6e6',

    globalNavBarBg: '#262626',

    // fonts
    baseFontColor: '#444',
    secondFontColor: '#777',

    // forms
    mandatoryFieldColor: '#a4030f',

    // leak
    leakColor: '#fbf3d5',
    leakBorderColor: '#eae3c7',

    // issues
    issueBgColor: '#f2dede',
    hotspotBgColor: '#eeeff4',
    issueLocationSelected: '#f4b1b0',
    issueLocationHighlighted: '#e1e1f2',
    conciseIssueRed: '#d18582',
    conciseIssueRedSelected: '#a4030f',

    // codeviewer
    lineCoverageRed: '#a4030f',
    lineCoverageGreen: '#b4dd78',

    // alerts
    warningIconColor: '#e2bf41',

    alertBorderError: '#f4b1b0',
    alertBackgroundError: '#f2dede',
    alertTextError: '#862422',
    alertIconError: '#a4030f',

    alertBorderWarning: '#faebcc',
    alertBackgroundWarning: '#fcf8e3',
    alertTextWarning: '#6f4f17',
    alertIconWarning: '#db781a',

    alertBorderSuccess: '#d6e9c6',
    alertBackgroundSuccess: '#dff0d8',
    alertTextSuccess: '#215821',
    alertIconSuccess: '#6d9867',

    alertBorderInfo: '#b1dff3',
    alertBackgroundInfo: '#d9edf7',
    alertTextInfo: '#0e516f',
    alertIconInfo: '#0271b9'
  },

  sizes: {
    gridSize: `${grid}px`,

    baseFontSize: '13px',
    verySmallFontSize: '10px',
    smallFontSize: '12px',
    mediumFontSize: '14px',
    bigFontSize: '16px',
    hugeFontSize: '24px',

    hugeControlHeight: `${5 * grid}px`,
    largeControlHeight: `${4 * grid}px`,
    controlHeight: `${3 * grid}px`,
    smallControlHeight: `${2.5 * grid}px`,
    tinyControlHeight: `${2 * grid}px`,

    globalNavHeight: `${6 * grid}px`,

    globalNavContentHeight: `${4 * grid}px`,

    maxPageWidth: '1320px',
    minPageWidth: '1080px',
    pagePadding: '20px'
  },

  rawSizes: {
    grid,
    globalNavHeightRaw: 6 * grid,
    globalNavContentHeightRaw: 4 * grid,
    contextNavHeightRaw: 9 * grid
  },

  fonts: {
    baseFontFamily: "'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif",
    systemFontFamily:
      "-apple-system,'BlinkMacSystemFont','Segoe UI','Helvetica','Arial',sans-serif",
    sonarcloudFontFamily:
      "Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif"
  },

  // z-index
  // =======
  //    1 -  100  for page elements (e.g. sidebars, panels)
  //  101 -  500  for generic page fixed elements (e.g. navigation, workspace)
  //  501 - 3000  for page ui elements
  // 3001 - 8000  for generic ui elements (e.g. dropdowns, tooltips)
  zIndexes: {
    // common
    aboveNormalZIndex: '3',
    normalZIndex: '2',
    belowNormalZIndex: '1',

    // ui elements
    pageMainZIndex: '50',

    tooltipZIndex: '8000',

    dropdownMenuZIndex: '7500',

    processContainerZIndex: '7000',

    modalZIndex: '6001',
    modalOverlayZIndex: '6000',

    popupZIndex: '5000'
  },

  others: {
    defaultShadow: '0 6px 12px rgba(0, 0, 0, 0.175)'
  }
};
