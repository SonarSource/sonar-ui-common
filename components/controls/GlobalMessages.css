/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.processes-container {
  position: fixed;
  z-index: var(--processContainerZIndex);
  top: 0;
  left: 50%;
  width: 350px;
  margin-left: -175px;
}

.process-spinner {
  position: relative;
  padding: 0 10px;
  line-height: var(--controlHeight);
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  background-color: #f0e8ac;
  text-align: center;
  opacity: 0;
  transition: all 0.2s ease;
}

.process-spinner.shown {
  opacity: 1;
}

.process-spinner + .process-spinner {
  margin-top: 5px;
  border-radius: 3px;
}

.process-spinner-failed {
  padding-right: 30px;
  background-color: var(--red);
  color: #ffffff;
}

.process-spinner-success {
  padding-right: 30px;
  background-color: var(--green);
  color: #ffffff;
}

.process-spinner-close {
  position: absolute;
  top: 2px;
  right: 2px;
}

.process-spinner-close:hover path,
.process-spinner-close:focus path {
  fill: var(--red) !important;
}
