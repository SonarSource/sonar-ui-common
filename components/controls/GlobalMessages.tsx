/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { keyframes } from '@emotion/core';
import * as React from 'react';
import { cutLongWords } from '../../helpers/path';
import { styled, themeGet, themeSize } from '../theme';
import { ClearButton } from './buttons';

interface Message {
  id: string;
  level: 'ERROR' | 'SUCCESS';
  message: string;
}

export interface GlobalMessagesProps {
  closeGlobalMessage: (id: string) => void;
  messages: Message[];
}

export default function GlobalMessages({ closeGlobalMessage, messages }: GlobalMessagesProps) {
  if (messages.length === 0) {
    return null;
  }

  return (
    <MessagesContainer>
      {messages.map((message) => (
        <GlobalMessage closeGlobalMessage={closeGlobalMessage} key={message.id} message={message} />
      ))}
    </MessagesContainer>
  );
}

const MessagesContainer = styled.div`
  position: fixed;
  z-index: ${themeGet('zIndexes', 'processContainerZIndex')};
  top: 0;
  left: 50%;
  width: 350px;
  margin-left: -175px;
`;

export class GlobalMessage extends React.PureComponent<{
  closeGlobalMessage: (id: string) => void;
  message: Message;
}> {
  handleClose = () => {
    this.props.closeGlobalMessage(this.props.message.id);
  };

  render() {
    const { message } = this.props;
    return (
      <Message data-test={`global-message__${message.level}`} level={message.level}>
        {cutLongWords(message.message)}
        <CloseButton
          className="button-small"
          color="#fff"
          level={message.level}
          onClick={this.handleClose}
        />
      </Message>
    );
  }
}

const appearAnim = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

const Message = styled.div<Pick<Message, 'level'>>`
  position: relative;
  padding: 0 30px 0 10px;
  line-height: ${themeSize('controlHeight')};
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  color: #ffffff;
  background-color: ${({ level, theme }) =>
    level === 'SUCCESS' ? theme.colors.green : theme.colors.red};
  text-align: center;
  opacity: 0;
  animation: ${appearAnim} 0.2s ease forwards;

  & + & {
    margin-top: calc(${themeSize('gridSize')} / 2);
    border-radius: 3px;
  }
`;

const CloseButton = styled(ClearButton)<Pick<Message, 'level'>>`
  position: absolute;
  top: calc(${themeSize('gridSize')} / 4);
  right: calc(${themeSize('gridSize')} / 4);

  &:hover svg,
  &:focus svg {
    color: ${({ level, theme }) => (level === 'SUCCESS' ? theme.colors.green : theme.colors.red)};
  }
`;
