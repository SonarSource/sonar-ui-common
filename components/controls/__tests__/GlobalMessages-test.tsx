/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import { matchers } from 'jest-emotion';
import * as React from 'react';
// mockedTheme is in fact the default theme
// eslint-disable-next-line jest/no-mocks-import
import { mockedTheme } from '../../__mocks__/mockedTheme';
import GlobalMessages, { GlobalMessagesProps } from '../GlobalMessages';

expect.extend(matchers);

it('should not render when no message', () => {
  expect(shallowRender({ messages: [] }).type()).toBeNull();
});

it('should render correctly with a message', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
  expect(
    wrapper
      .find('GlobalMessage')
      .first()
      .dive()
  ).toMatchSnapshot();
});

it('should render with correct css', () => {
  const wrapper = shallowRender();
  expect(wrapper.render()).toMatchSnapshot();
  expect(
    wrapper
      .find('GlobalMessage')
      .first()
      .render()
  ).toHaveStyleRule('background-color', mockedTheme.colors.red);

  expect(
    wrapper
      .find('GlobalMessage')
      .last()
      .render()
  ).toHaveStyleRule('background-color', mockedTheme.colors.green);
});

function shallowRender(props: Partial<GlobalMessagesProps> = {}) {
  return shallow(
    <GlobalMessages
      closeGlobalMessage={jest.fn()}
      messages={[
        { id: '1', level: 'ERROR', message: 'Test' },
        { id: '2', level: 'SUCCESS', message: 'Test 2' }
      ]}
      {...props}
    />
  );
}
