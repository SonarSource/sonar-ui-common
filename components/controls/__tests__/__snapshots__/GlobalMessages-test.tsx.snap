// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with a message 1`] = `
<Styled(div)>
  <GlobalMessage
    closeGlobalMessage={[MockFunction]}
    key="1"
    message={
      Object {
        "id": "1",
        "level": "ERROR",
        "message": "Test",
      }
    }
  />
  <GlobalMessage
    closeGlobalMessage={[MockFunction]}
    key="2"
    message={
      Object {
        "id": "2",
        "level": "SUCCESS",
        "message": "Test 2",
      }
    }
  />
</Styled(div)>
`;

exports[`should render correctly with a message 2`] = `
<Styled(div)
  data-test="global-message__ERROR"
  level="ERROR"
  role="alert"
>
  Test
  <Styled(ClearButton)
    className="button-small"
    color="#fff"
    level="ERROR"
    onClick={[Function]}
  />
</Styled(div)>
`;

exports[`should render correctly with a message 3`] = `
<Styled(div)
  data-test="global-message__SUCCESS"
  level="SUCCESS"
  role="status"
>
  Test 2
  <Styled(ClearButton)
    className="button-small"
    color="#fff"
    level="SUCCESS"
    onClick={[Function]}
  />
</Styled(div)>
`;

exports[`should render with correct css 1`] = `
@keyframes animation-0 {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes animation-0 {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.emotion-4 {
  position: fixed;
  z-index: 7000;
  top: 0;
  left: 50%;
  width: 350px;
  margin-left: -175px;
}

.emotion-1 {
  position: relative;
  padding: 0 30px 0 10px;
  line-height: 24px;
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  color: #ffffff;
  background-color: #d4333f;
  text-align: center;
  opacity: 0;
  -webkit-animation: animation-0 0.2s ease forwards;
  animation: animation-0 0.2s ease forwards;
}

.emotion-1 + .emotion-1 {
  margin-top: calc(8px / 2);
  border-radius: 3px;
}

.emotion-0 {
  position: absolute;
  top: calc(8px / 4);
  right: calc(8px / 4);
}

.emotion-0:hover svg,
.emotion-0:focus svg {
  color: #d4333f;
}

.emotion-3 {
  position: relative;
  padding: 0 30px 0 10px;
  line-height: 24px;
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  color: #ffffff;
  background-color: #00aa00;
  text-align: center;
  opacity: 0;
  -webkit-animation: animation-0 0.2s ease forwards;
  animation: animation-0 0.2s ease forwards;
}

.emotion-3 + .emotion-3 {
  margin-top: calc(8px / 2);
  border-radius: 3px;
}

.emotion-2 {
  position: absolute;
  top: calc(8px / 4);
  right: calc(8px / 4);
}

.emotion-2:hover svg,
.emotion-2:focus svg {
  color: #00aa00;
}

<div
  class="emotion-4"
>
  <div
    class="emotion-1"
    data-test="global-message__ERROR"
    role="alert"
  >
    Test
    <button
      class="button button-small emotion-0 button-icon"
      level="ERROR"
      style="color:#fff"
      type="button"
    >
      <svg
        height="16"
        space="preserve"
        style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:1.41421"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
        xlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M14 4.242L11.758 2l-3.76 3.76L4.242 2 2 4.242l3.756 3.756L2 11.758 4.242 14l3.756-3.76 3.76 3.76L14 11.758l-3.76-3.76L14 4.242z"
          style="fill:currentColor"
        />
      </svg>
    </button>
  </div>
  <div
    class="emotion-3"
    data-test="global-message__SUCCESS"
    role="status"
  >
    Test 2
    <button
      class="button button-small emotion-2 button-icon"
      level="SUCCESS"
      style="color:#fff"
      type="button"
    >
      <svg
        height="16"
        space="preserve"
        style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:1.41421"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
        xlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M14 4.242L11.758 2l-3.76 3.76L4.242 2 2 4.242l3.756 3.756L2 11.758 4.242 14l3.756-3.76 3.76 3.76L14 11.758l-3.76-3.76L14 4.242z"
          style="fill:currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
